### PR TITLE
Implement `sendrawtransaction` RPC

### LINKED ISSUE #3146: sendrawtransaction JSON-RPC method
## Motivation

`lightwalletd` uses the `sendrawtransaction` JSON-RPC method.

### Design Decisions

`zcashd` puts the transaction in its mempool, so Zebra should do the same thing.

What should we do if we think the transaction is invalid?
- log it at warning level
- If Zebra rejects a transaction that `zcashd` accepts, that's a bug in Zebra we need to fix

Zebra doesn't implement pending transactions in the mempool, so the transaction can't depend on another mempool transaction.

### Required Fields

The method is documented here: https://zcash.github.io/rpc/sendrawtransaction.html

`lightwalletd` only uses `hexstring` (raw transaction data) as an argument: https://github.com/zcash/lightwalletd/blob/master/frontend/service.go#L302

Field list:
```
"hex"             (string) The transaction hash in hex - ignored by `lightwalletd`?
```

### Tasks

- Implement the RPC method 
- Add RPC acceptance tests to CI
- Test that the RPC method works with `lightwalletd`
- Test with transactions that get split across multiple TCP packets - use a max-sized transaction

We need to test with large transactions because our JSON protocol fixes operate on each packet.

## API Reference

We plan to use [jsonrpc_core](https://docs.rs/jsonrpc-core/18.0.0/jsonrpc_core/) with:
* [jsonrpc_http_server](https://github.com/paritytech/jsonrpc/tree/master/http)
* [jsonrpc-derive](https://github.com/paritytech/jsonrpc/tree/master/derive)

### Example Code

Here are examples of:
* [declaring RPC methods as Rust functions](https://github.com/paritytech/jsonrpc#basic-usage-with-derive)

## Out of Scope

`lightwalletd` seems to do some weird error handling, but this is out of scope for Zebra: https://github.com/zcash/lightwalletd/blob/master/frontend/service.go#L308

### PR DESCRIPTION
**Pull request by jvff**
## Motivation

In order to support `lightwalletd`, Zebra needs to provide certain remote procedure calls. One of them, `sendrawtransaction`, allows sending a signed transaction to the network.

### Specifications

`zcashd`'s documentation for this RPC: https://zcash.github.io/rpc/sendrawtransaction.html

## Solution

Implement a `send_raw_transaction` method that handles the `sendrawtransaction` RPC request. The received transaction, if it can be properly deserialized from the input hexadecimal string, is sent to the mempool service.

The `RpcImpl` type keeps a handle to the mempool service (a generic type parameter) in order to send the raw transaction to it.

This closes #3146.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

## Follow Up Work

1. I implemented the error handling in a way that I thought was sensible, but I'm now thinking if we should check `zcashd`'s error handling and reproduce it exactly.
2. The current implementation stores the transaction in the mempool, but that doesn't guarantee that the transaction is gossiped to other nodes. Maybe this RPC call should also force the transaction to be gossiped to at least a few nodes?

### CODE REVIEW

**Comment by teor2345** [comment]
> * I implemented the error handling in a way that I thought was sensible, but I'm now thinking if we should check `zcashd`'s error handling and reproduce it exactly.

We don't expect errors during normal operation, so this seems like enough for now.

> * The current implementation stores the transaction in the mempool, but that doesn't guarantee that the transaction is gossiped to other nodes. Maybe this RPC call should also force the transaction to be gossiped to at least a few nodes?

This could be a premature workaround. In the design review, we decided to just send transactions to the mempool in the lightwalletd minimum viable product.

If we have any transaction propagation issues, they could be bugs in transaction validation, the mempool, or network code. Once we know what they are, we'll have enough info to triage them.

File: zebra-rpc/src/methods.rs
Comment (teor2345) [inline zebra-rpc/src/methods.rs]: Missing documentation: `sendrawtransaction` has an optional `allowhighfees` argument, can you note that it is never used by `lightwalletd`?

File: zebra-rpc/src/methods.rs
Comment (teor2345) [inline zebra-rpc/src/methods.rs]: Question: do we need to say "signed" here, or is the extra information confusing?

So far, every transaction in Zebra has signatures and proofs.

File: zebra-rpc/src/methods.rs
Comment (teor2345) [inline zebra-rpc/src/methods.rs:189]: Question: would this be better as a typed structure?

Would that need more `serde` derivations in `zebra-chain`? And do we want to make those changes?

```suggestion
pub struct SentTransactionHash(transaction::Hash);
```

**Review by teor2345 — COMMENTED** [review]
Looks great!

I have a few minor design questions, but I'd be happy to merge it as it is.

The ticket for the missing send transaction lightwalletd integration test is #3512.

File: zebra-rpc/src/methods.rs
Comment (dconnolly) [reply zebra-rpc/src/methods.rs]: Yeah a valid transaction has  (sometimes multiple) signatures, if it doesn't it won't be parsed correctly and rejected

File: zebra-rpc/src/methods.rs
Comment (dconnolly) [inline zebra-rpc/src/methods.rs]:
```suggestion
                .map_err(|_| Error::invalid_params("raw transaction is structurally invalid"))?;
```

File: zebra-rpc/src/methods.rs
Comment (dconnolly) [reply zebra-rpc/src/methods.rs:189]: I like this change, looks like `transaction::Hash` is ready to go:
image

File: zebra-rpc/src/methods/tests/prop.rs
Comment (dconnolly) [inline zebra-rpc/src/methods/tests/prop.rs:18]: _nice_

File: zebra-rpc/src/methods/tests/prop.rs
Comment (dconnolly) [inline zebra-rpc/src/methods/tests/prop.rs:185]: nice

**Comment by dconnolly** [comment]
Some small suggestions but this looks great, the tests are awesome

File: zebra-rpc/src/methods.rs
Comment (teor2345) [reply zebra-rpc/src/methods.rs:189]: From the snapshot tests, I think the default `serde` array serialization is an integer list of byte values, not a hexadecimal string.

So I'm happy to go with whatever is simpler here.

But if we can make a change that simplifies the remaining 9 RPCs, let's do that.

File: zebra-rpc/src/methods.rs
Comment (oxarbitrage) [inline zebra-rpc/src/methods.rs]: I think we should agree in a doc format for all methods. I proposed a format in https://github.com/ZcashFoundation/zebra/discussions/3638 and it is the one already being used by `getinfo`. Please check https://github.com/ZcashFoundation/zebra/discussions/3638 and let me know how do you see it.

File: zebra-rpc/src/methods.rs
Comment (oxarbitrage) [inline zebra-rpc/src/methods.rs]: Are we sure this Is this the only call that will use the mempool service ?

**Review by oxarbitrage — COMMENTED** [review]
The tests are great. I think we should try to merge this one ASAP as it will be conflicting with several others. I only have the question i made about RPC methods documentation.

File: zebra-rpc/src/methods.rs
Comment (jvff) [reply zebra-rpc/src/methods.rs]: Oops, I hadn't seen that discussion! I proposed a few changes there, but I'm not sure how to proceed:

1. Wait until we finish the discussion to update this PR and merge it.
2. Merge it with a non-standard documentation and open a separate PR to fix it later.
3. Change the documentation to follow `getinfo`'s documentation and open a separate PR to fix both of them later only if there are changes to the proposal.

I'm leaning towards 3, but if anyone else has any preferences or alternatives please let me know :sweat_smile:

File: zebra-rpc/src/methods.rs
Comment (jvff) [reply zebra-rpc/src/methods.rs]: Done.

File: zebra-rpc/src/methods.rs
Comment (jvff) [reply zebra-rpc/src/methods.rs]: I'm not 100% sure :thinking: We might also use it in `getrawtransaction`, but I think we can update this documentation in the PR for that RPC method if needed :sweat_smile:

File: zebra-rpc/src/methods.rs
Comment (jvff) [reply zebra-rpc/src/methods.rs:189]: I had an idea on how to improve it, but since it expanded the scope a bit and risks making things a bit more complex, I opened a separate PR for it (#3706). If that's a reasonable approach we can merge it into this PR first, otherwise we can merge this PR first and either tweak that PR more to merge it or just close/replace it.

File: zebra-rpc/src/methods.rs
Comment (teor2345) [reply zebra-rpc/src/methods.rs]: There are 3 calls that use the mempool service:
* getrawmempool
* getrawtransaction
* sendrawtransaction

See the data flow diagram here:
https://docs.google.com/drawings/d/1F3qCpIhzqhCCfWot5YEvsIt9gixvlwZd2TQNqxcOyBk/edit

File: zebra-rpc/src/methods.rs
Comment (oxarbitrage) [reply zebra-rpc/src/methods.rs]: I think your suggestions in the discussion should be included in the final format. However, i agree we can use the current `getinfo` and open a ticket to change them all after we decide, should be easy. I also have another one that will need updates at https://github.com/ZcashFoundation/zebra/pull/3707

File: zebra-rpc/src/methods.rs
Comment (oxarbitrage) [inline zebra-rpc/src/methods.rs]:
```suggestion
```

File: zebra-rpc/src/methods.rs
Comment (oxarbitrage) [reply zebra-rpc/src/methods.rs]: With that in mind, i added a suggestion at https://github.com/ZcashFoundation/zebra/pull/3685/files#r818188945

File: zebra-rpc/src/methods.rs
Comment (teor2345) [reply zebra-rpc/src/methods.rs:189]: It think it's more complex to start with, but it allows the compiler to check that we're using the correct types.

File: zebra-rpc/src/methods.rs
Comment (jvff) [reply zebra-rpc/src/methods.rs]: Applied, thanks!

**Review by teor2345 — DISMISSED** [review]
Looks good, let's go!

**Comment by jvff** [comment]
I had to resolve a merge conflict before it could be merged :sweat_smile:

**Review by dconnolly — APPROVED** [review]

**Pull request merged**
